frontend/src/presentation/components/QuestionFilters.tsx

```diff
@@ -0,0 +1,181 @@
+import React, { useState, useMemo, useCallback } from 'react';
+import { Select, Input, Tag, Dropdown, Button } from 'antd';
+import { DownOutlined } from '@ant-design/icons';
+import styles from './QuestionFilters.module.css';
+
+const { Option } = Select;
+const { CheckableTag } = Tag;
+
+interface QuestionFiltersProps {
+  allCategories: string[];
+  onFiltersChange: (filters: {
+    selectedDifficulty: string;
+    selectedCategories: string[];
+    searchTerm: string;
+  }) => void;
+}
+
+export const QuestionFilters: React.FC<QuestionFiltersProps> = ({
+  allCategories,
+  onFiltersChange,
+}) => {
+  const [selectedDifficulty, setSelectedDifficulty] = useState<string>('All');
+  const [selectedCategories, setSelectedCategories] = useState<string[]>([]);
+  const [categorySearchTerm, setCategorySearchTerm] = useState('');
+  const [searchTerm, setSearchTerm] = useState('');
+
+  const triggerFiltersChange = useCallback(
+    (filters: {
+      selectedDifficulty: string;
+      selectedCategories: string[];
+      searchTerm: string;
+    }) => {
+      if (onFiltersChange) {
+        onFiltersChange(filters);
+      }
+    },
+    [onFiltersChange]
+  );
+
+  const handleSearch = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const value = e.target.value;
+      setSearchTerm(value);
+      triggerFiltersChange({
+        selectedDifficulty,
+        selectedCategories,
+        searchTerm: value,
+      });
+    },
+    [selectedDifficulty, selectedCategories, triggerFiltersChange]
+  );
+
+  const handleDifficultyChange = useCallback(
+    (value: string) => {
+      setSelectedDifficulty(value);
+      triggerFiltersChange({
+        selectedDifficulty: value,
+        selectedCategories,
+        searchTerm,
+      });
+    },
+    [selectedCategories, searchTerm, triggerFiltersChange]
+  );
+
+  const handleCategorySearch = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      setCategorySearchTerm(e.target.value);
+    },
+    []
+  );
+
+  const handleCategoryTagChange = useCallback(
+    (category: string, checked: boolean) => {
+      const nextSelectedCategories = checked
+        ? [...selectedCategories, category]
+        : selectedCategories.filter((c) => c !== category);
```

Comment:
Can consider optional chaining, just to be extra safe :)
```suggestion
        : selectedCategories?.filter((c) => c !== category);
```

Reply:
Why not, let's just put it in.